### ISSUE TITLE
Wrong type narrowing in class  methods

### ISSUE THREAD

**Issue by nivekithan** [microsoft/TypeScript#45026]
# Bug Report

### 🔎 Search Terms

- Control Flow in class methods
- Wrong type narrowing in class methods

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about this bug

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/FAYwNghgzlAECiAPCBbADmAprA3sWBsA7gPYBOAJrAFywBEEdsAPvQEZOt0h0Dcw+QiBIA7KABcyAVxDjyACgCUuQYULiAFgEsoAOlKVYAXnqNVsAL4C1IDRBEBzTABUSAISUq1azTv3kqEwByNiDzK3NbeydXAGFPPG9CczUtADNYeQBCeV89A0CjEwY6RWVEpO9NMhIiWBFMOvgyGrJ5JlKUwgjK2DzdKMcXdyV+XutKgHpJ2ABJcVgAZQ0SKTAqERIF6trYTBbyLoJ0zJz+guMi9lLyiaSe7uALIA

### 💻 Code

```ts
class Example {
    word : "a" | "b" | "c";

    constructor() {
        this.word = "a"
    }

    changeToB() {
        this.word = 'b'
    }

    changeToC() {
        
        if (!(this.word === "a")) {
            throw new Error(" ")
        }
    


        this.changeToB();
        

        // It Should not throw error, the type of this.word should be "a" | "b" | "c" 
        if (!(this.word === "b")) {

        }

    }
}
```
### 🙁 Actual behavior

Right now after narrowing `this.word` type to `"a"` , `typescript` assumes that the `type`  is constant for entire block . It does not seem to understand that a method call could change the `value` of  a `this.word`  as  a result `type` wont be same.

### 🙂 Expected behavior

After it sees a method call , the typescript should forgot the previously remembered type of `this.word`.

If I really want typescript to remember the type then they I can write code like this

```ts
class Example {
    word : "a" | "b" | "c";

    constructor() {
        this.word = "a"
    }

    changeToB() {
        this.word = 'b'
    }

    changeToC() {
        const word = this.word;

        if (!(word === "a")) {
            throw new Error(" ")
        }

        // Shows Error: Correct behaviour
        if (word === "b") {

        }

    }
}
```

**Comment by MartinJohns** [comment]
See #9998.

**Comment by nivekithan** [comment]
Thanks for clarification I will be closing the issue.
